File: PrefixSum/523.continuous-subarray-sum.py
from typing import List


# @leet start
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        preSum = 0
        m = {0: -1}
        for i, v in enumerate(nums):
            preSum += v
            remain = preSum % k
            if remain in m:
                if i - m[remain] >= 2:
                    return True
            else:
                m[remain] = i

        return False


# @leet end

File: PrefixSum/523.continuous-subarray-sum.go
package leetcode

// @leet start
func checkSubarraySum(nums []int, k int) bool {
	var (
		preSum int
		m      = map[int]int{0: -1}
	)
	for i, v := range nums {
		preSum += v
		remain := preSum % k
		if preIdx, ok := m[remain]; ok {
			if i-preIdx >= 2 {
				return true
			}
		} else {
			m[remain] = i
		}
		// NOTE: m[remain] = i // need max length, avoid overwrite
	}
	return false
}

// @leet end

